src/node_trace_events.cc

```diff
@@ -0,0 +1,164 @@
+#include "node_internals.h"
+#include "tracing/agent.h"
+
+namespace node {
+
+using v8::Context;
+using v8::FunctionCallbackInfo;
+using v8::Int32;
+using v8::Local;
+using v8::Object;
+using v8::Value;
+
+// The tracing APIs require category groups to be pointers to long-lived
+// strings. Those strings are stored here.
+static std::unordered_set<std::string> categoryGroups;
+
+// Gets a pointer to the category-enabled flags for a tracing category group,
+// if tracing is enabled for it.
+static const uint8_t* GetCategoryGroupEnabled(const char* category_group) {
+    if (category_group == nullptr) return nullptr;
+
+    return TRACE_EVENT_API_GET_CATEGORY_GROUP_ENABLED(category_group);
+}
+
+static const char* GetCategoryGroup(Environment* env,
+                                    const Local<Value> categoryValue) {
+  if (!categoryValue->IsString()) {
+    env->ThrowTypeError("trace_events category must be a string.");
+    return nullptr;
+  }
+
+  Utf8Value category(env->isolate(), categoryValue);
+  // If the value already exists in the set, insertion.first will point
+  // to the existing value. Thus, this will maintain a long lived pointer
+  // to the category c-string.
+  auto insertion = categoryGroups.insert(category.out());
+
+  // The returned insertion is a pair whose first item is the object that was
+  // inserted or that blocked the insertion and second item is a boolean
+  // indicating whether it was inserted.
+  return insertion.first->c_str();
+}
+
+static inline char GetPhase(Environment* env, const Local<Value> arg) {
+  if (!arg->IsNumber()) {
+    env->ThrowTypeError("trace_events type must be a number.");
```

Comment:
Can the type checks be moved up to JS and use `internal/errors` instead?

Reply:
I'm not sure. We need to make sure that `GetPhase` doesn't have unexpected behavior for unexpected input, right? We are definitely doing this in a lot of other places, so if it is an issue we need to find a more general solution.

We can certainly create an extra abstraction layer as `internal/trace_events.js` and then repeat the type checking there.

Reply:
Remember, this is not a public interface. If this error ever throws it is because there is a bug in node core. The API stability of the message shouldn't really be a concern.

Reply:
Perhaps these should be `CHECK`s then?